Fix `contao:user:list` with empty database

AlexanderWillner opened the pull request:
When executing `contao-console contao:user:list` on an empty database we'd receive the following error otherwise:
`Return value of Contao\CoreBundle\Command\UserListCommand::getUsers() must be an instance of Contao\Model\Collection, null returned`.

With this change instead we receive a nice warning: ` ! [NOTE] No accounts found.`

Commit:
Fixing 'contao-console contao:user:list' on empty database.

When executing 'contao-console contao:user:list' on an empty database
we'd receive the following error otherwise:
'Return value of Contao\CoreBundle\Command\UserListCommand::getUsers() must be an instance of Contao\Model\Collection, null returned'

aschempp commented:
nice catch! I think we should use an array instead of returning a collection, then this problem would have happened in the first place. This also does not fix the problem when checking for admin only accounts, as that might still return `null`.

Commit:
added missing import

Commit:
avoid crash if folder does not exist - e.g. assets/preview

Commit:
getUsers now returns an array

AlexanderWillner commented:
> nice catch! I think we should use an array instead of returning a collection, then this problem would have happened in the first place. 

I've refactored the code in d2d4236 - have you had something like this in mind?

> This also does not fix the problem when checking for admin only accounts, as that might still return `null`.

right - also fixed in d2d4236

leofeyer commented:
Why not allow `null` as return type?

```php
private function getUsers(bool $onlyAdmins = false): ?Collection
```

AlexanderWillner commented:
> Why not allow `null` as return type?
> 
> ```
> private function getUsers(bool $onlyAdmins = false): ?Collection
> ```

That's an alternative. In this case we'd need to change line 69 from

```
if (0 === $users->count()) {
```

to something like

```
if (! $users || 0 === $users->count()) {
```

as well. Otherwise we get a the error `Call to a member function count() on null`.

Commit:
Revert "added missing import"

This reverts commit 35508b405f2d7b503f91d5846ae8834ebf9784b7.

Commit:
Revert "avoid crash if folder does not exist - e.g. assets/preview"

This reverts commit 25f3b1d67371652e99027f9c1e07a60f02a13f07.

Commit:
Revert "getUsers now returns an array"

This reverts commit d2d42368c9779ffda3ff9d50be1cf2bd7bf78c99.

Commit:
Revert "Fixing 'contao-console contao:user:list' on empty database."

This reverts commit edbc94fa8d6a7a0fa9155ae41e540f9fb5884d3b.

Commit:
solution with using null as return type

AlexanderWillner committed:
Merge branch 'contao:4.13' into 4.13

leofeyer commented on core-bundle/src/Command/UserListCommand.php:
```diff
@@ -133,10 +133,10 @@ static function (string $field) use ($user) {
     {
         if ([] === $columns) {
-            return $users->fetchAll();
+            return $users ? $users->fetchAll() : [];
```
There should be an early return at the beginning of the method, because line 144 might still trigger the error.

```php
private function formatJson(?Collection $users, array $columns): array
{
    if (!$users) {
        return [];
    }

    if ([] === $columns) {
        return $users->fetchAll();
    }

    // …
```

Commit:
added early return

AlexanderWillner committed:
Merge branch '4.13' into 4.13

aschempp reviewed (approved):
Thanks @AlexanderWillner and sorry for having to refactor this twice 😅

leofeyer commented:
Thank you @AlexanderWillner.

AlexanderWillner commented:
> Thanks @AlexanderWillner and sorry for having to refactor this twice 😅

refactoring is fun ;)
